internal/mode/static/state/graph/policies.go

```diff
@@ -0,0 +1,310 @@
+	Ancestor *PolicyAncestor
+	// TargetRef is the resource that the Policy targets.
+	TargetRef PolicyTargetRef
+	// Conditions holds the conditions for the Policy.
+	// These conditions apply to the entire Policy.
+	// The conditions in the Ancestor apply only to the Policy in regard to the Ancestor.
+	Conditions []conditions.Condition
+	// Valid indicates whether the Policy is valid.
+	Valid bool
+}
+
+// PolicyAncestor represents an ancestor of a Policy.
+type PolicyAncestor struct {
+	// Ancestor is the ancestor object.
+	Ancestor v1alpha2.ParentReference
+	// Conditions contains the list of conditions of the Policy in relation to the ancestor.
+	Conditions []conditions.Condition
+}
+
+// PolicyAncestorRef contains the identifying information of the ancestor.
+type PolicyAncestorRef struct {
+	// Nsname is the NamespacedName of the Ancestor.
+	Nsname types.NamespacedName
+	// Kind is the Kind of the Ancestor.
+	Kind v1.Kind
+	// Group is the Group of the Ancestor.
+	Group v1.Group
+}
+
+// PolicyTargetRef represents the object that the Policy is targeting.
+type PolicyTargetRef struct {
+	// Kind is the Kind of the object.
+	Kind v1.Kind
+	// Group is the Group of the object.
+	Group v1.Group
+	// Nsname is the NamespacedName of the object.
+	Nsname types.NamespacedName
+}
+
+// PolicyKey is a unique identifier for an NGF Policy.
+type PolicyKey struct {
+	// Nsname is the NamespacedName of the Policy.
+	NsName types.NamespacedName
+	// GVK is the GroupVersionKind of the Policy.
+	GVK schema.GroupVersionKind
+}
+
+const (
+	gatewayGroupKind = v1.GroupName + "/" + kinds.Gateway
+	hrGroupKind      = v1.GroupName + "/" + kinds.HTTPRoute
+	grpcGroupKind    = v1.GroupName + "/" + kinds.GRPCRoute
+)
+
+// attachPolicies attaches the graph's processed policies to the resources they target. It modifies the graph in place.
+func (g *Graph) attachPolicies(ctlrName string) {
+	if g.Gateway == nil {
+		return
+	}
+
+	for _, policy := range g.NGFPolicies {
+		ref := policy.TargetRef
+
+		switch ref.Kind {
+		case kinds.Gateway:
+			attachPolicyToGateway(policy, g.Gateway, g.IgnoredGateways, ctlrName)
+		case kinds.HTTPRoute, kinds.GRPCRoute:
+			route, exists := g.Routes[routeKeyForKind(ref.Kind, ref.Nsname)]
+			if !exists {
+				continue
+			}
+
+			attachPolicyToRoute(policy, route, ctlrName)
+		}
+	}
+}
+
+func attachPolicyToRoute(policy *Policy, route *L7Route, ctlrName string) {
+	kind := v1.Kind(kinds.HTTPRoute)
+	if route.RouteType == RouteTypeGRPC {
+		kind = kinds.GRPCRoute
+	}
+
+	routeNsName := types.NamespacedName{Namespace: route.Source.GetNamespace(), Name: route.Source.GetName()}
+
+	ancestor := &PolicyAncestor{
+		Ancestor: createParentReference(v1.GroupName, kind, routeNsName),
+	}
+
+	curAncestorStatus := policy.Source.GetPolicyStatus().Ancestors
+	if ancestorsFull(curAncestorStatus, ancestor.Ancestor, ctlrName) {
+		return
+	}
+
+	policy.Ancestor = ancestor
+
+	if !route.Valid || !route.Attachable || len(route.ParentRefs) == 0 {
```

Comment:
If the route doesn't have any parentRefs, then it's not even considered in the graph. So this particular check may not be necessary.

Also, `TargetRefNotFound` is kind of a weird reason if the target ref does exist, but is invalid. But it looks like that's all the GW API has to offer right now.

Reply:
Another thing I noticed with my code (not sure if the problem exists in yours as well) is that if the route is deleted, the ancestor status in the policy still remains. This is due to the fact that we no longer consider the policy attached to anything, so we don't use it or update its status anymore.

Do we remove status from other resources if they are no longer relevant?

Reply:
> If the route doesn't have any parentRefs, then it's not even considered in the graph. So this particular check may not be necessary.

I added that check because of this line of code: https://github.com/nginxinc/nginx-gateway-fabric/blob/35b2aabbafac3d41443a3853a5f785c90a299d11/internal/mode/static/state/graph/httproute.go#L25.

I figured it was a possible edge case. Is it not?

Reply:
> Also, TargetRefNotFound is kind of a weird reason if the target ref does exist, but is invalid. But it looks like that's all the GW API has to offer right now.

Agreed, but the comment for the condition reason made me think it applies to this case:

> // PolicyReasonTargetNotFound is used with the "Accepted" condition when the
// policy is attached to an invalid target resource.

Reply:
> I figured it was a possible edge case. Is it not?

Ah yeah, I see the error case could apply. Nevermind.

> // PolicyReasonTargetNotFound is used with the "Accepted" condition when the
> // policy is attached to an invalid target resource.

yeah the wording does make it apply, even though the name doesn't match up

Reply:
> Another thing I noticed with my code (not sure if the problem exists in yours as well) is that if the route is deleted, the ancestor status in the policy still remains. This is due to the fact that we no longer consider the policy attached to anything, so we don't use it or update its status anymore.

I think this is known issue/limitation of our status implementation: https://github.com/nginxinc/nginx-gateway-fabric/issues/1015

It's definitely problematic and we should prioritize fixing it. 


> Do we remove status from other resources if they are no longer relevant?